server/util/networking/networking.go

```diff
@@ -346,6 +336,35 @@ func (a *HostNetAllocator) unlock(netIdx int) {
 //	# add a route in the root namespace so that traffic to 192.168.0.3 hits 10.0.0.2, the veth0 end of the pair
 //	$ sudo ip route add 192.168.0.3 via 10.0.0.2
 func SetupVethPair(ctx context.Context, netNamespace, vmIP string, vmIdx int) (_ func(context.Context) error, err error) {
+	// Keep a list of cleanup work to be done. For each resource we create in
+	// this func that needs to be explicitly cleaned up, we append a cleanup
+	// task to this list. Cleanup work is done in the reverse order in which
+	// it's added to this list (i.e., it's a stack).
+	var cleanupStack []func(ctx context.Context) error
+	cleanup := func(ctx context.Context) error {
+		ctx, cancel := background.ExtendContextForFinalization(ctx, networkingCleanupTimeout)
+		defer cancel()
+		// Pop and run cleanup funcs from the stack until empty.
+		for len(cleanupStack) > 0 {
+			f := cleanupStack[len(cleanupStack)-1]
+			cleanupStack = cleanupStack[:len(cleanupStack)-1]
+			if err := f(ctx); err != nil {
+				// Short-circuit on the first error. Note, this means we may not
+				// unlock the host IP, so fire a warning.
+				alert.UnexpectedEvent("vm_network_cleanup_failed", "VM networking cleanup failed. This might keep the VM host IP locked, so if there are a lot of these errors, VM networking may eventually stop functioning. Error: %s", err)
+				return err
+			}
+		}
+		return nil
+	}
+	// If we return an error from this func then we need to clean up any
+	// resources that were created before returning.
+	defer func() {
+		if err != nil {
```

Comment:
When `err` is shadowed - i.e. `hostEndpointNet, err := hostNetAllocator.Get(vmIdx)` - does this still work correctly?

Reply:
Yeah - in that example, `err` is not being declared (shadowed), it's being re-assigned. The declaration happens in the named return value.

I was on the fence about whether to do this vs using a wrapper func but I think ultimately the wrapper func is more error prone because we have to remember to `return cleanup, err` instead of `return nil, err` everywhere in this func.

Reply:
Ah okay. I wasn't sure if a new `err` variable was being created, but it seems like it won't, and only the value will get reassigned (https://go.dev/ref/spec#Short_variable_declarations)